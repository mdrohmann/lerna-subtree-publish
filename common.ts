import * as path from "path"
import { Subtrees, SubtreeConfig } from "./types"
import * as execa from "execa"
import * as fs from "fs"
import { promisify } from "util"
import * as color from "colour"

// a short shim that I use because shell.exec(..., {silent: true}) believes that
// it does not always return an ExecOutputReturnValue, so it needs an explicit
// cast.
export const getSubtrees = async (
  cwd: string = process.cwd()
): Promise<Subtrees> => {
  try {
    return JSON.parse(
      await promisify(fs.readFile)(path.join(cwd, "subtrees.json"), "utf8")
    )
  } catch (error) {
    throw new Error(
      `subtrees.json file does not exit, or could not be parsed:\n${error}`
    )
  }
}

export const getSubtreeConfig = async (
  treeName: string,
  cwd: string = process.cwd()
) => {
  const subtrees = await getSubtrees(cwd)
  if (subtrees.hasOwnProperty(treeName)) {
    return subtrees[treeName]
  } else {
    console.error(`Unknown subtree ${treeName}`)
    throw new Error(`Unknown subtree ${treeName}`)
  }
}

export const gitSubtreeSplit = async (
  config: SubtreeConfig,
  ref: string = "master",
  cwd: string = process.cwd()
) => {
  const res = execa.shell(
    `git subtree split --prefix=${config.localFolder} ${ref} --squash`,
    { cwd }
  )
  res.stdout.pipe(process.stdout)

  const result = await res

  if (result.code !== 0) {
    throw new Error(`Could not push to ${config.repository}\n:${res.stderr}`)
  }
  return result.stdout
}

export const gitSubtreeCmd = async (
  config: SubtreeConfig,
  command: string,
  treeName: string,
  cwd: string = process.cwd()
) => {
  const res = execa.shell(
    `git subtree ${command} --prefix=${config.localFolder} ${treeName} ${
      config.branch
    } --squash`,
    { cwd }
  )
  res.stdout.pipe(process.stdout)

  const result = await res

  if (result.code !== 0) {
    throw new Error(`Could not push to ${treeName}\n:${res.stderr}`)
  }
}

export interface LernaJson {
  subtreeInitScript?: string
}

export const getLernaJson = async (cwd: string = process.cwd()) => {
  const lernaJsonString = await promisify(fs.readFile)(
    path.join(cwd, "lerna.json"),
    "utf8"
  )
  return JSON.parse(lernaJsonString) as LernaJson
}

export const getSubtreePackageJson = async (
  c: SubtreeConfig,
  ref: string = "HEAD",
  cwd: string = process.cwd()
) => {
  const pkgJsonString = await execa.shell(
    `git show ${ref}:${c.localFolder}/package.json`,
    { cwd }
  )
  return JSON.parse(pkgJsonString.stdout)
}

export const gitFolderHashes = async (
  folder: string,
  num: number = 1,
  cwd: string = process.cwd()
) => {
  const currentFolderRes = await execa.shell(
    `git log -n ${num} --pretty=format:%h -- ${folder}`,
    { cwd }
  )
  if (currentFolderRes.code !== 0) {
    throw new Error(currentFolderRes.stderr)
  }
  const currentHash = currentFolderRes.stdout.trim()
  return currentHash.split("\n")
}

export const gitTagOfHash = async (
  hash: string,
  match?: string,
  cwd: string = process.cwd()
) => {
  const currentTagRes = await execa.shell(
    `git describe --abbrev=0 ${
      match === undefined ? "" : `--match='${match}*'`
    } ${hash}`,
    { cwd }
  )
  if (currentTagRes.code !== 0) {
    throw Error(
      `Error get tag at hash ${hash} with match ${match}*\n${
        currentTagRes.stderr
      }`
    )
  }
  return currentTagRes.stdout.trim()
}

export const gitTag = async (
  tag: string,
  message: string,
  ref: string = "HEAD",
  cwd: string = process.cwd()
) => {
  await execa.shell(`git tag -a ${tag} -m "${message}" ${ref}`, { cwd })
}

export const gitTagDelete = async (
  tag: string,
  cwd: string = process.cwd()
) => {
  await execa.shell(`git tag -d ${tag}`, { cwd })
}

export const gitInit = async (cwd: string = process.cwd()) => {
  await execa.shell(`git init`, { cwd })
}

export const gitAdd = async (path: string, cwd: string = process.cwd()) => {
  return execa.shell(`git add ${path}`, { cwd })
}

export const gitCommit = async (
  message: string,
  cwd: string = process.cwd()
) => {
  return execa.shell(`git commit -m"${message}"`, { cwd })
}

export const gitPush = async (
  upstream: string,
  ref: string,
  cwd: string = process.cwd()
) => {
  return execa.shell(`git push --follow-tags ${upstream} ${ref}`, { cwd })
}

export const gitHashOfTag = async (
  tag: string,
  cwd: string = process.cwd()
) => {
  const res = await execa.shell(`git rev-list -n 1 --pretty=format:%h ${tag}`, {
    cwd
  })

  if (res.code !== 0) {
    throw new Error(`Could not retrieve SHA of ref ${tag}: ${res.stderr}`)
  }
  return res.stdout
    .trim()
    .split("\n")
    .pop()
}

interface GitRuntimeOptions {
  cwd?: string
  gitDir?: string
}

export const parseGitRuntimeOptions = (rOptions: GitRuntimeOptions) => {
  const res: {
    cwd: string
    env?: { [variable: string]: string }
  } = {
    cwd: rOptions.cwd || process.cwd()
  }
  if (rOptions.gitDir !== undefined) {
    res["env"] = {
      ...process.env,
      GIT_DIR: rOptions.gitDir
    }
  }
  return res
}

export const gitHashOfCommitRef = async (
  ref: string,
  options: GitRuntimeOptions = {}
) => {
  const res = await execa.shell(
    `git rev-parse --short ${ref}`,
    parseGitRuntimeOptions(options)
  )

  if (res.code !== 0) {
    throw new Error(`Could not retrieve SHA of ref ${ref}: ${res.stderr}`)
  }
  return res.stdout.trim()
}

/**
 * return the git hash and tag of the latest publication of a package at a given
 * sub-tree
 * @param subtrees subtree definitions
 * @param treeName the sub-tree to look for
 */
export const lastPublishedTag = async (
  subtrees: Subtrees,
  treeName: string,
  cwd: string = process.cwd()
) => {
  const config = subtrees[treeName]
  const currentHash = (await gitFolderHashes(config.localFolder, 1, cwd))[0]

  const pkgName = (await getSubtreePackageJson(config, "HEAD", cwd)).name

  const currentTag = await (async () => {
    // first try to get a tag that resolves to pkgName@version
    try {
      const res = await gitTagOfHash(currentHash, pkgName, cwd)
      return res
    } catch (err) {
      return gitTagOfHash(currentHash, undefined, cwd)
    }
  })()

  const currentTagHash = await gitHashOfTag(currentTag, cwd)

  return {
    tag: currentTag,
    hash: currentTagHash,
    current: currentTagHash === currentHash
  }
}

export const gitChangesPending = async (cwd: string = process.cwd()) => {
  const res = await execa.shell("git status --porcelain --untracked=no", {
    cwd
  })
  if (res.code !== 0) {
    throw new Error(`could not check for pending changes`)
  }

  if (res.stdout !== "") {
    throw new Error(
      `Working tree has modifications. Sort them out first.\n${res.stdout}`
    )
  }
  return undefined
}

export const gitGetRemotes = async (cwd: string = process.cwd()) => {
  return (await execa.shell("git remote", { cwd })).stdout.split("\n")
}

export const gitRemoteAdd = async (
  treeName: string,
  repository: string,
  cwd: string = process.cwd()
) => execa.shell(`git remote add -f ${treeName} ${repository}`, { cwd })

export const gitSubtreeAdd = async (
  config: SubtreeConfig,
  treeName: string,
  cwd: string = process.cwd()
) => {
  const localFolderExists = await (async () => {
    try {
      await promisify(fs.readdir)(path.join(cwd, config.localFolder))
      return true
    } catch (error) {
      return false
    }
  })()
  const remotes = await gitGetRemotes(cwd)
  const remoteExists = remotes.indexOf(treeName) !== -1
  if (remoteExists) {
    console.log(`Remote ${color.yellow(treeName)} already exists`)
  } else {
    await gitRemoteAdd(treeName, config.repository, cwd)
  }
  if (localFolderExists) {
    console.log(
      `Local folder ${color.yellow(config.localFolder)} already exists`
    )
  } else {
    await execa
      .shell(`git fetch ${treeName}`, { cwd })
      .stdout.pipe(process.stdout)

    await execa.shell(
      `git subtree add --prefix=${config.localFolder} ${treeName} ${
        config.branch
      } --squash`,
      { cwd }
    )
  }
}

export const commandAll = async (
  handler: (config: SubtreeConfig, treeName: string) => Promise<void>,
  skipCurrentTrees: boolean = false,
  cwd: string = process.cwd()
) => {
  await gitChangesPending(cwd)
  const subtrees = await getSubtrees(cwd)
  await Promise.all(
    Object.keys(subtrees).map(async treeName => {
      if (
        skipCurrentTrees &&
        (await lastPublishedTag(subtrees, treeName, cwd)).current
      ) {
        console.log(
          `Skipping ${color.yellow(treeName)}: Sub-tree has not changed.`
        )
      } else {
        const config = subtrees[treeName]
        await handler(config, treeName)
      }
    })
  )
}

export const commandSingle = async (
  treeName: string,
  handler: (config: SubtreeConfig, treeName: string) => Promise<void>,
  cwd: string = process.cwd()
) => {
  await gitChangesPending(cwd)
  const config = await getSubtreeConfig(treeName, cwd)
  await handler(config, treeName)
}

export const getUpdatedPackages = async (cwd: string = process.cwd()) => {
  try {
    const res = await execa.shell("lerna updated --json", { cwd })
    const updates = res.stdout.toString().trim()
    if (updates === "") {
      return []
    }
    return JSON.parse(updates) as ReadonlyArray<{
      name: string
      version: string
      private: boolean
    }>
  } catch (error) {
    return []
  }
}
