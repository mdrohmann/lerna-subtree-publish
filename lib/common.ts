import * as child_process from "child_process"
export const embedRuntime = (
  command: string,
  subCommand: string,
  args: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => {
  child_process.execFileSync(command, [subCommand].concat(args), {
    stdio: "inherit",
    cwd
  })

  if (
    args.length === 1 &&
    (args[0] === "-h" ||
      args[0] === "-v" ||
      args[0] === "--help" ||
      args[0] === "--version")
  ) {
    process.exit(0)
  }
}
