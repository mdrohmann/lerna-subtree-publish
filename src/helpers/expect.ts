import * as execa from "execa"
import * as fs from "fs"
import * as path from "path"
import { promisify } from "util"

export const getGitDir = async (cwd: string) => {
  if (await promisify(fs.exists)(path.join(cwd, ".git"))) {
    return path.join(cwd, ".git")
  } else {
    return cwd
  }
}

export const expectableCommitTree = async (cwd: string, refs: string[]) => {
  return (await execa.shell(
    `git log --decorate --pretty=format:'commit %d' --graph ${refs.join(" ")}`,
    {
      cwd,
      env: {
        GIT_DIR: await getGitDir(cwd)
      }
    }
  )).stdout
}

export const expectableSubtrees = async (
  cwd: string,
  repositoryBase: string
) => {
  return (await promisify(fs.readFile)(
    path.join(cwd, "subtrees.json"),
    "utf8"
  )).replace(new RegExp(repositoryBase, "g"), "repository")
}
