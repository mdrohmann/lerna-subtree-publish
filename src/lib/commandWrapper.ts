import { SubtreeConfig } from "../types"
import { gitChangesPending } from "./git"
import { getSubtrees, getSubtreeConfig } from "./gitSubtree"
import { lastPublishedTag } from "./lastPublishedTag"
import * as color from "colour"

export const asyncExecution = async <I, O>(
  inputs: I[],
  handler: (input: I) => Promise<O>,
  parallel: boolean
): Promise<O[]> => {
  if (parallel) {
    return await Promise.all(inputs.map(handler))
  } else {
    let result: O[] = []
    for (let index = 0; index < inputs.length; ++index) {
      const element = await handler(inputs[index])
      result.push(element)
    }
    return result
  }
}

export const commandAll = async (
  handler: (config: SubtreeConfig, treeName: string) => Promise<void>,
  skipCurrentTrees: boolean = false,
  parallel: boolean = true,
  cwd: string = process.cwd()
) => {
  await gitChangesPending(cwd)
  const subtrees = await getSubtrees(cwd)
  await asyncExecution(
    Object.keys(subtrees),
    async treeName => {
      if (
        skipCurrentTrees &&
        (await lastPublishedTag(subtrees, treeName, cwd)).current
      ) {
        console.log(
          `Skipping ${color.yellow(treeName)}: Sub-tree has not changed.`
        )
      } else {
        const config = subtrees[treeName]
        await handler(config, treeName)
      }
    },
    parallel
  )
}

export const commandSingle = async (
  treeName: string,
  handler: (config: SubtreeConfig, treeName: string) => Promise<void>,
  cwd: string = process.cwd()
) => {
  await gitChangesPending(cwd)
  const config = await getSubtreeConfig(treeName, cwd)
  await handler(config, treeName)
}
