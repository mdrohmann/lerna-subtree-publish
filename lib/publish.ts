import {
  getUpdatedPackages,
  gitSubtreeCmd,
  commandAll,
  gitChangesPending,
  gitFolderHashes,
  gitTag,
  gitPush,
  gitTagDelete,
  gitSubtreeSplit
} from "../common"
import * as color from "colour"
import { gitHashOfCommitRef, getSubtreePackageJson } from "../common"
import { embedRuntime } from "./common"

const embedLernaPublish = (
  argv: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => embedRuntime("lerna", "publish", argv.slice(2), cwd)

export const lernaPublish = async (
  argv: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => {
  if (Object.keys(await getUpdatedPackages(cwd)).length === 0) {
    console.error("Nothing to publish!")
    return
  }

  // assert that there are no pending changes
  await gitChangesPending(cwd)

  // TODO: make sure we run the prepush hooks (probably lernaPublish will take
  // care of that...)
  /*
  // push all sub-trees that have changed
  await commandAll(
    async (c, tn) => {
      await gitSubtreeCmd(c, "push", tn, cwd)
    },
    true,
    cwd
  )
  */

  embedLernaPublish(argv, cwd)

  await commandAll(
    async (c, tn) => {
      // push sub-tree
      await gitSubtreeCmd(c, "push", tn, cwd)

      const lastTwoChanges = await gitFolderHashes(c.localFolder, 2, cwd)

      const headSha = await gitHashOfCommitRef("HEAD", { cwd })
      const lastChangeSha = await gitHashOfCommitRef(lastTwoChanges[0], { cwd })

      if (headSha === lastChangeSha) {
        const lastTwoVersions = await Promise.all(
          lastTwoChanges.map(
            async (ref: string) =>
              (await getSubtreePackageJson(c, ref, cwd)).version
          )
        )
        if (lastTwoVersions[0] !== lastTwoVersions[1]) {
          console.log(
            `We have a new version for ${color.green(tn)}: Tagging and pushing!`
          )
          const newVersion = lastTwoVersions[0]
          const newVersionTag = `v${newVersion}`

          const splitHash = await gitSubtreeSplit(c, "master", cwd)

          await gitTag(newVersionTag, `version newVersion`, splitHash, cwd)

          await gitPush(tn, newVersionTag, cwd)

          await gitTagDelete(newVersionTag, cwd)
        } else {
          console.log(`${color.yellow(tn)}: no version change: skip tagging`)
        }
      } else {
        console.log(`${color.yellow(tn)}: no new publication, skip tagging`)
      }
    },
    false,
    cwd
  )
}
