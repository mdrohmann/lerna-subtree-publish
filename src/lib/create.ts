import { embedRuntime } from "./common"
import * as path from "path"
import { SubtreeConfig } from "../types"
import * as fs from "fs"
import { promisify } from "util"
import * as execa from "execa"
import rimraf from "rimraf"
import mkdirp from "mkdirp-promise"
import {
  gitChangesPending,
  gitInit,
  gitRemoteAdd,
  gitPush,
  gitAdd,
  gitCommit,
  gitTag
} from "./git"
import { getSubtrees, gitSubtreeAdd, getSubtreePackageJson } from "./gitSubtree"
import { getLernaJson } from "./lerna"

const embedNpmInit = (
  argv: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => embedRuntime("npm", "init", argv, cwd)

export const parseRemotes = () => {}

export const lernaSplitArgs = (argv: ReadonlyArray<string>) => {
  if (argv.length < 3) {
    throw new Error(`You need to specify a package name`)
  }

  const repoIndex = argv.findIndex(x => x === "--repository")
  const repo = repoIndex === -1 ? undefined : argv[repoIndex + 1]
  const newArgv =
    repoIndex === -1
      ? argv.slice(3)
      : argv.slice(3, repoIndex).concat(argv.slice(repoIndex + 2))

  return {
    name: argv[3],
    repo,
    argv: newArgv
  }
}

export const lernaCreate = async (
  name: string,
  repository: string,
  initArgs: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => {
  // assert that there are no pending changes
  await gitChangesPending(cwd)

  const subtreesJsonExists = await promisify(fs.exists)(
    path.join(cwd, "subtrees.json")
  )

  const oldSubtrees = subtreesJsonExists ? await getSubtrees(cwd) : {}
  // Assert that name is not taken yet.
  if (oldSubtrees.hasOwnProperty(name)) {
    throw new Error(`A subtree with name ${name} already exists.`)
  }
  const localFolder = path.join("packages", name)
  const packagePath = path.join(cwd, localFolder)

  if (await promisify(fs.exists)(packagePath)) {
    throw new Error(`A package already exists under ${localFolder}`)
  }

  // ensure that packagePath exists
  await mkdirp(packagePath)

  // TODO: maybe we should allow to create the packages in other folders
  embedNpmInit(initArgs, packagePath)

  // make it private by default
  const packageJsonFile = path.join(packagePath, "package.json")
  const packageJson = JSON.parse(
    await promisify(fs.readFile)(packageJsonFile, "utf8")
  )
  await promisify(fs.writeFile)(
    packageJsonFile,
    JSON.stringify(
      {
        ...packageJson,
        private: true
      },
      null,
      2
    ),
    "utf8"
  )

  const config: SubtreeConfig = {
    localFolder,
    repository,
    branch: "master"
  }

  /* add the remote repository...
  console.log("npm init")
  await gitSubtreeAdd(config, name, cwd)
  */

  const lernaJson = await getLernaJson(cwd)
  if (lernaJson.subtreeInitScript !== undefined) {
    await execa.shell(`${lernaJson.subtreeInitScript}`, {
      cwd,
      env: {
        ...process.env,
        REPOSITORY: repository,
        PACKAGE_NAME: name
      }
    })
  }

  // We cannot just push changes to create repo... We need to use subtree add at
  // some point..., so we have to
  const result = await (async () => {
    try {
      // * initialize a dummy git repo here
      await gitInit(packagePath)
      // * make an initial commit
      await execa.shell(`git add .`, { cwd: packagePath })
      await execa.shell('git commit -m"initial commit"', { cwd: packagePath })
      // * add origin and push
      await gitRemoteAdd("origin", repository, packagePath)
      const res = await gitPush("origin", "HEAD", true, packagePath)
      return {
        okay: res.code === 0,
        error: res.stderr
      }
    } catch (error) {
      return {
        okay: false,
        error
      }
    }
  })()
  // clean up the git repository.
  await promisify(rimraf)(path.join(packagePath, ".git"))

  if (!result.okay) {
    throw new Error(
      `Error pushing to new repository ${repository}.\nYou can re-run this command later, when you know how to fix the error:\n${
        result.error
      }`
    )
  }

  // Here, we can be sure that the remote repo is pushed, so we:
  // * delete the package folder
  await promisify(rimraf)(packagePath)
  // * add and it as a subtree repo
  // * update subtrees.json
  const newSubtrees = {
    ...oldSubtrees,
    [name]: config
  }
  await promisify(fs.writeFile)(
    path.join(cwd, "subtrees.json"),
    JSON.stringify(newSubtrees, null, 2),
    "utf8"
  )
  await gitAdd("subtrees.json", cwd)
  await gitCommit(`add subtree package ${name}`, cwd)

  // * add the package back
  await gitSubtreeAdd(config, name, cwd)

  // TODO: I believe that this should go into gitSubtreeAdd...
  const version = (await getSubtreePackageJson(config, "HEAD", cwd)).version
  const tag = `${name}@${version}`
  await gitTag(tag, `Add package ${name} with version ${version}`, "HEAD", cwd)
  await gitPush("origin", "HEAD", true, cwd)
}
