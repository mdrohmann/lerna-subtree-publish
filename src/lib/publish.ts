import * as color from "colour"
import { embedRuntime } from "./common"
import {
  gitChangesPending,
  gitFolderHashes,
  gitHashOfCommitRef,
  gitTag,
  gitPush,
  gitTagDelete
} from "./git"
import { getUpdatedPackages } from "./lerna"
import { commandAll } from "./commandWrapper"
import { gitSubtreeSplit, getSubtreePackageJson } from "./gitSubtree"

const embedLernaPublish = (
  argv: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => embedRuntime("lerna", "publish", argv.slice(2), cwd)

export const lernaPublish = async (
  argv: ReadonlyArray<string>,
  cwd: string = process.cwd()
) => {
  if (Object.keys(await getUpdatedPackages(cwd)).length === 0) {
    console.error("Nothing to publish!")
    return
  }

  // assert that there are no pending changes
  await gitChangesPending(cwd)

  // TODO: make sure we run the prepush hooks (probably lernaPublish will take
  // care of that...)
  /*
  // push all sub-trees that have changed
  await commandAll(
    async (c, tn) => {
      await gitSubtreeCmd(c, "push", tn, cwd)
    },
    true,
    cwd
  )
  */

  embedLernaPublish(argv, cwd)

  await commandAll(
    async (c, tn) => {
      const lastTwoChanges = await gitFolderHashes(c.localFolder, 2, cwd)

      const headSha = await gitHashOfCommitRef("HEAD", { cwd })
      const lastChangeSha = await gitHashOfCommitRef(lastTwoChanges[0], { cwd })

      if (headSha === lastChangeSha) {
        // this package is affected by the publication, so split it up and
        // eventually push it, but first check if we have to tag it too.
        const splitHash = await gitSubtreeSplit(c, "", cwd)

        const lastTwoVersions = await Promise.all(
          lastTwoChanges.map(
            async (ref: string) =>
              (await getSubtreePackageJson(c, ref, cwd)).version
          )
        )
        if (lastTwoVersions[0] !== lastTwoVersions[1]) {
          console.log(
            `We have a new version for ${color.green(
              tn
            )}: Tag it before pushing!`
          )
          const newVersion = lastTwoVersions[0]
          const newVersionTag = `v${newVersion}`

          await gitTag(newVersionTag, `version newVersion`, splitHash, cwd)

          await gitPush(tn, `${splitHash}:refs/heads/${c.branch}`, true, cwd)

          await gitTagDelete(newVersionTag, cwd)
        } else {
          console.log(`${color.yellow(tn)}: no version change: skip tagging`)
          // we still need to push, this can happen if our version dependencies have changed...
          // TODO: Test this code branch
          await gitPush(tn, `${splitHash}:refs/heads/${c.branch}`, true, cwd)
        }
      } else {
        console.log(`${color.yellow(tn)}: no new publication, skip tagging`)
      }
    },
    false,
    cwd
  )
}
