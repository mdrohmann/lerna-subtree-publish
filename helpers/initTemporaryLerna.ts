import * as tmp from "tmp"
import { promisify } from "util"
import * as fs from "fs"
import * as path from "path"
import * as execa from "execa"
import { SubtreeConfig } from "types"
import { gitTag, gitRemoteAdd, gitPush } from "../common"
import { lernaCreate } from "../lib/create"

export const LERNA_ROOT_NAME = "lerna-tmp-root"

export const makeSubtrees = (
  repositoryBase: string,
  packageNames: ReadonlyArray<string>
) =>
  packageNames.reduce<{
    [name: string]: SubtreeConfig
  }>((p, n) => {
    p[n] = {
      localFolder: path.join("packages", n),
      branch: "master",
      repository: path.join(repositoryBase, n)
    }
    return p
  }, {})

export const initTemporaryDirectory = () => {
  return new Promise<{ path: string; cleanupCallback: () => void }>(
    (resolve, reject) => {
      tmp.dir((err, path, cleanupCallback) => {
        if (err) {
          return reject(err)
        }
        return resolve({
          path,
          cleanupCallback
        })
      })
    }
  )
}

export const initLernaJson = async (directory: string) => {
  await promisify(fs.writeFile)(
    path.join(directory, "lerna.json"),
    JSON.stringify({
      lerna: "2.11.0",
      packages: ["packages/*"],
      version: "independent"
    }),
    "utf8"
  )
}

export const initPackageJson = async (
  directory: string,
  name: string = "lerna-tmp-example"
) => {
  await promisify(fs.writeFile)(
    path.join(directory, "package.json"),
    JSON.stringify({
      version: "1.0.0",
      name,
      private: true
    })
  )
}

export const initSubPackage = async (directory: string) => {
  try {
    await promisify(fs.mkdir)(directory)
  } catch (err) {}
  await initPackageJson(directory, path.basename(directory))
}

export interface TemporaryDirectories {
  lernaBase: string
  path: string
  repositoryBase: string
}

export const initTemporaryLerna = async (
  numPackets: number
): Promise<TemporaryDirectories> => {
  const tempDir = await initTemporaryDirectory()

  const baseDir = path.join(tempDir.path, "lerna")

  fs.mkdirSync(baseDir)

  // init lerna.json
  await initLernaJson(baseDir)

  // init package.json
  await initPackageJson(baseDir)

  fs.mkdirSync(path.join(baseDir, "packages"))

  const packageIndices = [...Array(numPackets).keys()].map(x => x + 1)

  // set up temporary repositories
  const repositoryBase = path.join(tempDir.path, "remotes")
  fs.mkdirSync(repositoryBase)

  await initTemporaryBareRepository(repositoryBase, LERNA_ROOT_NAME)
  await Promise.all(
    packageIndices.map(p =>
      initTemporaryBareRepository(repositoryBase, `p${p}`)
    )
  )

  // make initial git commit
  await initializeAndPushLernaRepository(baseDir, repositoryBase)

  try {
    // init sub-packages (one after the other)
    for (let i = 0; i < packageIndices.length; ++i) {
      const p = packageIndices[i]
      await lernaCreate(
        `p${p}`,
        path.join(repositoryBase, `p${p}`),
        ["--yes"],
        baseDir
      )
    }
  } catch (err) {
    console.error("error creating packages", err)
  }

  /*
  // write subtrees.json
  const packages = packageIndices.map(p => `p${p}`)
  const subtrees = makeSubtrees(repositoryBase, packages)

  fs.writeFileSync(
    path.join(baseDir, "subtrees.json"),
    JSON.stringify(subtrees)
  )
  */

  return {
    path: tempDir.path,
    repositoryBase,
    lernaBase: baseDir
  }
}

export const initTemporaryBareRepository = async (
  baseDir: string,
  name: string
) => {
  const repositoryPath = path.join(baseDir, name)
  await promisify(fs.mkdir)(repositoryPath)
  return execa.shell(`git init --bare`, { cwd: repositoryPath })
}

export const commitEverythingInPath = async (
  directory: string,
  message: string
) => {
  await execa.shell(`git add *`, { cwd: directory })
  return execa.shell(`git commit -m"${message}"`, { cwd: directory })
}

export const initializeAndPushLernaRepository = async (
  lernaBase: string,
  repositoryBase: string
) => {
  await execa.shell("git init", { cwd: lernaBase })
  await commitEverythingInPath(lernaBase, "initial")
  // tag it
  await gitTag("v1", "version 1", "HEAD", lernaBase)
  // add origin
  await gitRemoteAdd(
    "origin",
    path.join(repositoryBase, LERNA_ROOT_NAME),
    lernaBase
  )
  // and push
  await gitPush("origin", "HEAD", lernaBase)
}
