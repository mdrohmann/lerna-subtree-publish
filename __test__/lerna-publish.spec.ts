import {
  initTemporaryLerna,
  commitEverythingInPath,
  LERNA_ROOT_NAME
} from "../helpers/initTemporaryLerna"
import * as path from "path"
import * as fs from "fs"
import rimraf from "rimraf"
import { gitSubtreeAdd, commandAll } from "../common"
import { gitHashOfCommitRef } from "../common"
import { TemporaryDirectories } from "../helpers/initTemporaryLerna"
import { promisify } from "util"
import { expectableCommitTree } from "../helpers/expect"
import { lernaPublish } from "../lib/publish"

describe("lerna-publish", () => {
  let directories: TemporaryDirectories

  beforeAll(async done => {
    directories = await initTemporaryLerna(2)
    done()
  })

  afterAll(async done => {
    try {
      await promisify(rimraf)(directories.path)
    } catch (err) {
      console.error("what is going on here? ", err)
    }
    done()
  })

  it("should initialize the remote repositories correctly", async done => {
    await commandAll(
      async (c, tn) => gitSubtreeAdd(c, tn, directories.lernaBase),
      false,
      directories.lernaBase
    )
    const packagesHead = await Promise.all(
      [1, 2].map(p =>
        gitHashOfCommitRef(`remotes/p${p}/master`, {
          cwd: directories.lernaBase
        })
      )
    )

    const packagesRemoteHead = await Promise.all(
      [1, 2].map(p =>
        gitHashOfCommitRef("HEAD", {
          gitDir: path.join(directories.path, "remotes", `p${p}`)
        })
      )
    )

    expect(packagesHead).toEqual(packagesRemoteHead)

    expect(
      await expectableCommitTree(directories.lernaBase, ["master"])
    ).toMatchSnapshot()

    /*
    const lernaHead = await gitHashOfCommitRef("HEAD", {
      cwd: directories.lernaBase
    })
    const lernaRemoteHead = await gitHashOfCommitRef("HEAD", {
      gitDir: path.join(directories.path, "remotes", LERNA_ROOT_NAME)
    })
    expect(lernaHead).toEqual(lernaRemoteHead)
    */

    done()
  })

  it("should do nothing if nothing has changed", async done => {
    const treeBefore = await expectableCommitTree(directories.lernaBase, [
      "master",
      "p1/master",
      "p2/master"
    ])
    await lernaPublish(
      ["", "", "--cd-version", "major", "--yes"],
      directories.lernaBase
    )
    const treeAfter = await expectableCommitTree(directories.lernaBase, [
      "master",
      "p1/master",
      "p2/master"
    ])
    expect(treeBefore).toEqual(treeAfter)
    done()
  })

  it("should do the lerna publish work and tag remote sub-tree repository", async done => {
    fs.writeFileSync(
      path.join(directories.lernaBase, "packages", "p1", "a"),
      "test",
      "utf8"
    )
    await commitEverythingInPath(directories.lernaBase, "a added")
    await lernaPublish(
      ["", "", "--cd-version", "major", "--yes"],
      directories.lernaBase
    )

    expect(
      await expectableCommitTree(directories.lernaBase, ["master", "p1/master"])
    ).toMatchSnapshot()

    expect(
      await expectableCommitTree(path.join(directories.repositoryBase, "p1"), [
        "master"
      ])
    ).toMatchSnapshot()

    expect(
      await expectableCommitTree(
        path.join(directories.repositoryBase, LERNA_ROOT_NAME),
        ["master"]
      )
    ).toMatchSnapshot()
    done()
    // make sure that gitsbt commands are only run on sub-repos that have changed since publication
    // make sure prepush and precomit hooks are run only once.
  })
})
