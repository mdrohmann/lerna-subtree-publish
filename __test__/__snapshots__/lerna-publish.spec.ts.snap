// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lerna-publish should do the lerna publish work and tag remote sub-tree repository 1`] = `
"* commit  (HEAD -> master, tag: p1@2.0.0, origin/master)
* commit 
*   commit  (tag: p2@1.0.0)
|\\\\  
| * commit 
* commit 
*   commit  (tag: p1@1.0.0)
|\\\\  
| * commit 
* commit 
* commit  (tag: v1)
* commit  (p1/master)
* commit 
* commit "
`;

exports[`lerna-publish should do the lerna publish work and tag remote sub-tree repository 2`] = `
"* commit  (HEAD -> master, tag: v2.0.0)
* commit 
* commit "
`;

exports[`lerna-publish should do the lerna publish work and tag remote sub-tree repository 3`] = `
"* commit  (HEAD -> master, tag: p1@2.0.0)
* commit 
*   commit  (tag: p2@1.0.0)
|\\\\  
| * commit 
* commit 
*   commit  (tag: p1@1.0.0)
|\\\\  
| * commit 
* commit 
* commit  (tag: v1)"
`;

exports[`lerna-publish should initialize the remote repositories correctly 1`] = `
"*   commit  (HEAD -> master, tag: p2@1.0.0, origin/master)
|\\\\  
| * commit 
* commit 
*   commit  (tag: p1@1.0.0)
|\\\\  
| * commit 
* commit 
* commit  (tag: v1)"
`;
